Support `ALTER TABLE ADD PRIMARY KEY` statements

frapa opened the pull request:
This PR implements `ALTER TABLE table ADD PRIMARY KEY (col1, ...);` statements, as per #57.

I suggest reviewing by commit, as I tried to make them thematic and self-contained. However, the first commits basically do changes needed for the last 2-3 commits, so please have a look at the final ones if something's unclear.

Let me know if there is anything I can improve.

frapa committed:
Add tests for adding primary key

frapa committed:
Add GetColumnIndices() method to BoundConstraint + refactor code to use it.

Taking advantage of polymorphism to replace
switch statements in a couple places.

I also had to refactor UniqueBoundConstraint
to use PhysicalIndexes instead of LogicalIndexes,
in a similar way to all other BoundConstraints.

frapa committed:
Rename some conflicting indices in for loops according to contribution guideline

frapa committed:
Decouple TransformConstraint function from PGListCell to reuse it later for ALTER TABLE statements

frapa committed:
Handle ALTER TABLE ADD PRIMARY KEY in transformer

frapa committed:
Move AddDataTableIndex to the DataTable class

I renamed it AddConstraintIndex.

By the name, this method belongs there,
and I need it later in the DataTable class
constructor to handle addition of unique indices.

frapa committed:
Make BindConstraint method reusable

frapa committed:
Handle constraint creation in TableEntry

frapa committed:
Add unique index creation to DataTable constructor.

Now, when a new unique index is created and a modified
instance of the storage is made, the DataTable will
automatically add an ART index and index existing
rows, throwing an error if the uniqueness is not
satisfied.

frapa committed:
Return pointer to index from TableListIndex on addition

This makes code that needs to access the index
after the addition to the table simpler.

frapa committed:
Fix formatting

Tishj commented:
I'm not sure this is the best time to add this feature, as there is a pending PR to rework binding of Indexes (#11867)

Tishj commented on src/catalog/catalog_entry/duck_table_entry.cpp:
```diff
@@ -754,6 +726,51 @@ unique_ptr<CatalogEntry> DuckTableEntry::DropForeignKeyConstraint(ClientContext
+
+	switch (info.constraint->type) {
+	case ConstraintType::UNIQUE: {
+		auto unique = reinterpret_cast<UniqueConstraint *>(info.constraint.get());
```
We have `info.constraint->Cast<UniqueConstraint>()` for this purpose

frapa replied:
I removed all appearances of this for `*Constraint` classes from the entire codebase.

Tishj commented on src/catalog/catalog_entry/duck_table_entry.cpp:
```diff
@@ -629,21 +600,22 @@ unique_ptr<CatalogEntry> DuckTableEntry::ChangeColumnType(ClientContext &context
 			} else if (bfk.info.type == ForeignKeyType::FK_TYPE_SELF_REFERENCE_TABLE) {
-				for (idx_t i = 0; i < bfk.info.fk_keys.size(); i++) {
-					key_set.insert(bfk.info.fk_keys[i]);
+				for (idx_t key_idx = 0; key_idx < bfk.info.fk_keys.size(); key_idx++) {
```
Since this is looping through all items anyways and the index isn't required inside the loop, I have a feeling this can be a `for (auto &key : bfk.info.fk_keys)` loop instead which is cleaner in my opinion

Tishj replied:
I think it's fine to make that change regardless, if you have to touch it anyways

frapa replied:
I changed this, actually I replaced the entire loop with an insert using iterators, hope that's the cleanest.

Tishj commented on src/planner/binder/statement/bind_create_table.cpp:
```diff
@@ -142,50 +58,120 @@ vector<unique_ptr<BoundConstraint>> Binder::BindNewConstraints(vector<unique_ptr
+	switch (constraint.type) {
+	case ConstraintType::CHECK: {
+		unique_ptr<BoundConstraint> bound_constraint = make_uniq<BoundCheckConstraint>();
+		auto check_constraint = reinterpret_cast<BoundCheckConstraint *>(bound_constraint.get());
```
`Cast<BoundCheckConstraint>`

frapa replied:
Will do this everywhere, not sure how I missed it.

Tishj commented on src/catalog/catalog_entry/duck_table_entry.cpp:
```diff
@@ -754,6 +726,51 @@ unique_ptr<CatalogEntry> DuckTableEntry::DropForeignKeyConstraint(ClientContext
+void DuckTableEntry::EnsureNoPrimaryKey() {
+	for (auto &constraint : constraints) {
+		if (constraint->type == ConstraintType::UNIQUE &&
+		    reinterpret_cast<UniqueConstraint *>(constraint.get())->is_primary_key) {
```
`Cast<UniqueConstraint>`

Tishj commented on src/catalog/catalog_entry/duck_table_entry.cpp:
```diff
@@ -13,74 +13,35 @@
-	vector<PhysicalIndex> indexes;
+vector<const ColumnDefinition *> GetUniqueConstraintKeys(const ColumnList &columns,
+                                                         const UniqueConstraint &constraint) {
+	vector<const ColumnDefinition *> index_columns;
```
We prefer either `optional_ptr` if it *has* to be a pointer (i.e it needs to change at some point or can be null)
In this instance I think this can be `vector<reference<ColumnDefinition>>` communicating better that this is never null and will not change

frapa replied:
This also should be solved everywhere.

Tishj commented on src/catalog/catalog_entry/duck_table_entry.cpp:
```diff
@@ -131,10 +93,16 @@ DuckTableEntry::DuckTableEntry(Catalog &catalog, SchemaCatalogEntry &schema, Bou
 				if (bfk.info.type == ForeignKeyType::FK_TYPE_FOREIGN_KEY_TABLE ||
 				    bfk.info.type == ForeignKeyType::FK_TYPE_SELF_REFERENCE_TABLE) {
 
+					vector<const ColumnDefinition *> index_columns;
```
Same comment as earlier, this does not need to be a pointer

Tishj commented on src/catalog/catalog_entry/duck_table_entry.cpp:
```diff
@@ -754,6 +726,51 @@ unique_ptr<CatalogEntry> DuckTableEntry::DropForeignKeyConstraint(ClientContext
+	// We need to create a new physical table with a new constraint and possibly a new
+	// unique index. This procedure is also responsible for checking that the existing
+	// data satisfies the new constraint.
+	auto bound_constraint = binder->BindConstraint(*info.constraint.get(), create_info->table, create_info->columns);
```
`*info.constraint`

Using `.get()` gives a raw pointer and evades the null check, causing this to segfault instead of throwing an InternalException if `constraint` were to be null

frapa replied:
I also replaced all occurrences of this with de-referencing.

Tishj commented on src/catalog/catalog_entry/duck_table_entry.cpp:
```diff
@@ -560,8 +531,8 @@ unique_ptr<CatalogEntry> DuckTableEntry::SetNotNull(ClientContext &context, SetN
-	auto new_storage = make_shared_ptr<DataTable>(
-	    context, *storage, make_uniq<BoundNotNullConstraint>(columns.LogicalToPhysical(LogicalIndex(not_null_idx))));
+	auto bound_constraint = make_uniq<BoundNotNullConstraint>(columns.LogicalToPhysical(LogicalIndex(not_null_idx)));
+	auto new_storage = make_shared_ptr<DataTable>(context, *storage, bound_constraint.get());
```
`*bound_constraint`

Tishj commented on src/catalog/catalog_entry/duck_table_entry.cpp:
```diff
@@ -754,6 +726,51 @@ unique_ptr<CatalogEntry> DuckTableEntry::DropForeignKeyConstraint(ClientContext
+	auto bound_constraint = binder->BindConstraint(*info.constraint.get(), create_info->table, create_info->columns);
+	auto bound_create_info = binder->BindCreateTableInfo(std::move(create_info), schema);
+
+	auto new_storage = make_shared_ptr<DataTable>(context, *storage, bound_constraint.get());
```
`*bound_constraint`

Tishj commented on src/storage/data_table.cpp:
```diff
@@ -131,18 +142,44 @@ DataTable::DataTable(ClientContext &context, DataTable &parent, idx_t removed_co
 
-	// Verify the new constraint against current persistent/local data
-	VerifyNewConstraint(context, parent, constraint.get());
+		vector<const ColumnDefinition *> index_columns;
```
`vector<reference<ColumnDefinition>> index_columns`

Tishj commented on src/storage/data_table.cpp:
```diff
@@ -1312,4 +1349,66 @@ vector<ColumnSegmentInfo> DataTable::GetColumnSegmentInfo() {
+	return info->indexes.AddIndex(std::move(art));
+}
+
+void DataTable::IndexExistingData(ClientContext &context, Index *index) {
```
`Index &index`

Tishj commented on src/storage/serialization/serialize_parse_info.cpp:
```diff
@@ -482,4 +485,15 @@ unique_ptr<ParseInfo> VacuumInfo::Deserialize(Deserializer &deserializer) {
 	return std::move(result);
 }
 
+void AddConstraintInfo::Serialize(Serializer &serializer) const {
```
These methods are generated by `scripts/generate_serialization.py` using `parse_info.json`

frapa replied:
I noticed too late :-). I will change that.

frapa replied:
Fixed.

Tishj commented on src/storage/table_index_list.cpp:
```diff
@@ -9,10 +9,11 @@
 	D_ASSERT(index);
 	lock_guard<mutex> lock(indexes_lock);
 	indexes.push_back(std::move(index));
+	return indexes.at(indexes.size() - 1).get();
```
this should return a `Index &`
Easiest way is probably to do:
```
auto &result = *index;
indexes.push_back(std::move(index));
return result;
```

frapa replied:
Fixed.

frapa commented:
Thanks for the quick review!

Regarding #11867, I guess that would affect the part about indexes, which should affect only small part of the code from this PR (the `DataTable` constructor). Review of the rest should be safe.

frapa committed:
Remove all reinterpret_casts for constraint classes

frapa committed:
Solve reference vs pointer comment

frapa commented on src/storage/data_table.cpp:
```diff
@@ -131,18 +142,44 @@ DataTable::DataTable(ClientContext &context, DataTable &parent, idx_t removed_co
 
 // Alter column to add new constraint
-DataTable::DataTable(ClientContext &context, DataTable &parent, unique_ptr<BoundConstraint> constraint)
+DataTable::DataTable(ClientContext &context, DataTable &parent, BoundConstraint *constraint)
```
I replaced all pointers with references in all methods I added.

frapa commented:
I fixed all comments, let me know if there is anything else I can change!

frapa committed:
Fix formatting

frapa committed:
AddConstraintInfo to represent ALTER TABLE ADD CONSTRAINT statements

frapa committed:
Generate enums

frapa committed:
Do not verify UNIQUE constraints

This is not necessary because
creating a unique index
automatically checks the uniqueness.

frapa committed:
Adapt code after change to index binding

frapa committed:
Add tests for transaction management and fix bug

frapa commented:
This PR is now ready for review, I updated the code to the master branch and fixed a few additional things:

- I fixed the code after the index binding PR #11867 (basically just cast to bound index).
- I added a few tests and fixes for transactions (there was a bug where indexes were retained even after rollback, caused by missing cloning of a vector).

taniabogatsch commented:
Hi @frapa, thanks for all your work on this new feature! 🎉  @Tishj asked me to have a go at reviewing your changes. As I'll be picking this PR up, it might help me if you could close all of the comments that you've resolved. Also, as this is a new feature, it makes sense to rebase the PR against the `feature` branch.

taniabogatsch commented on test/sql/alter/alter_col/test_add_primary_key_transaction.test:
Additionally, there are important scenarios not yet covered in the tests in this PR. What happens if you are scanning the table to build your index while another transactions alters the table and commits its changes? Or drops the table? You can write tests for this by using the `concurrentloop` functionality. Have a look at this test `test/sql/parallelism/interquery/concurrent_appends.test`. Prior to the `concurrentloop` functionality, we were writing these tests in dedicated `.cpp` files. I.e., for the (older) index functionality, you can have a look at this file: `test/sql/parallelism/interquery/test_concurrent_index.cpp`.

taniabogatsch replied:
We still need this test.

taniabogatsch commented on test/sql/alter/alter_col/test_add_primary_key.test:
Any time you add a PK constraint, or alter the data in a table after adding the PK constraint, it can help to use an index scan to additionally verify the row ids in the ART. Currently, your tests only cover that the ART *keys* are valid, which we use for constraint checking. The data in the ART is not tested. We 'only' use it for index scans, but we cannot break them.

taniabogatsch replied:
We should expand a few of the existing tests by adding more data and triggering the index joins. Testing different key types would also be beneficial, as the current tests use mostly integers.

taniabogatsch reviewed (changes requested):
Hi @frapa! Thanks for picking up this long-standing feature request. The PR is going in a great direction. I've added my comments, most importantly I've requested a lot of additional tests (around concurrency).

Tishj commented on test/sql/alter/alter_col/test_add_primary_key_transaction.test:
Some stress testing using `concurrentloop` and `statement maybe` could also be good to add
see for example `third_party/duckdb/test/sql/catalog/dependencies/test_concurrent_alter.test`

> Multiple transactions add the same primary key constraint, but only one can commit successfully. All others need to throw an exception on COMMIT

I would say that this also needs a `statement maybe`, this will most likely throw a TransactionException, write-write conflict

taniabogatsch replied:
Resolved. Most tests have been added; I'll add a separate comment about a WAL test.

frapa committed:
Fix indexing of existing data + minor fixes

frapa committed:
Replace duplicated helper with single method.

frapa commented:
I will add more tests in the next few days, for now I addressed the requested changed and fixed the problem with indexing.

frapa committed:
Fix formatting

Tishj commented:
Thank you for the changes!
You should target the branch to `feature` as well so only your changes show up in the "files changed", this can be done at the top of the page 👍

frapa committed:
Adding more tests

frapa committed:
Adding more tests

frapa committed:
Fix linting error

Tishj commented on src/storage/data_table.cpp:
```diff
@@ -30,6 +30,17 @@
+IndexStorageInfo GetIndexInfo(const IndexConstraintType &constraint_type, string &table_name,
+                              const vector<reference<const ColumnDefinition>> &columns) {
+	auto constraint_name = EnumUtil::ToString(constraint_type);
+	string concat_columns;
+	for (const auto &col : columns) {
+		concat_columns += "_";
+		concat_columns += col.get().Name();
+	}
```
```
postgres=# CREATE TABLE pk_tbl (i int, j int);
CREATE TABLE
postgres=# create index pk_tbl_pkey on pk_tbl(j);
CREATE INDEX
postgres=# ALTER TABLE pk_tbl ADD PRIMARY KEY (i);
ALTER TABLE
postgres=# \d pk_tbl
               Table "public.pk_tbl"
 Column |  Type   | Collation | Nullable | Default 
--------+---------+-----------+----------+---------
 i      | integer |           | not null | 
 j      | integer |           |          | 
Indexes:
    "pk_tbl_pkey1" PRIMARY KEY, btree (i)
    "pk_tbl_pkey" btree (j)
```

taniabogatsch commented on test/sql/alter/alter_col/test_add_primary_key_transaction.test:
```diff
@@ -0,0 +1,140 @@
+statement ok
+COMMIT
+
+# Scenario #6: Multiple transaction try to add primary key
+statement ok
+CREATE TABLE test6 (i INTEGER, j INTEGER)
+
+statement ok tran1
+BEGIN TRANSACTION
+
+statement ok tran1
+ALTER TABLE test6 ADD PRIMARY KEY (j)
+
+statement ok tran2
+BEGIN TRANSACTION
+
+statement error tran2
+ALTER TABLE test6 ADD PRIMARY KEY (j)
+----
+TransactionContext Error: Catalog write-write conflict on alter with "test6"
```
Can we slightly expand this scenario like so:

```sql
statement ok tran3
BEGIN TRANSACTION

statement ok tran1
ROLLBACK

statement ok tran2
ROLLBACK

statement ok tran3
ALTER TABLE test6 ADD PRIMARY KEY (j)

statement ok tran3
COMMIT

statement error
INSERT INTO test6 VALUES (1, 1), (1, 1)
----
PRIMARY KEY or UNIQUE constraint violated
```

taniabogatsch replied:
Done.

taniabogatsch commented on test/sql/alter/alter_col/test_add_primary_key_database.test:
Let's add another file where we test `ALTER TABLE xx ADD PRIMARY KEY(..)` with the WAL. Any changes must persist when restarting, even if we did not checkpoint.
```sql
statement ok
PRAGMA disable_checkpoint_on_shutdown

statement ok
PRAGMA wal_autocheckpoint='1TB';
```

taniabogatsch replied:
File created.

taniabogatsch commented on test/sql/alter/alter_col/test_add_primary_key.test:
```diff
@@ -0,0 +1,206 @@
+statement ok
+CREATE TABLE test3 (i INTEGER PRIMARY KEY, j INTEGER)
+
+statement error
+ALTER TABLE test3 ADD PRIMARY KEY (i, j)
+----
+Catalog Error: table "test3" can have only one primary key, and already has PRIMARY KEY(i).
```
Let's also test these cases:
```sql
statement error
ALTER TABLE test3 ADD PRIMARY KEY (i)
----
error msg

statement error
ALTER TABLE test3 ADD PRIMARY KEY (j)
----
error msg
```

taniabogatsch commented on test/sql/alter/alter_col/test_add_primary_key.test:
```diff
@@ -0,0 +1,206 @@
+----
+Table with name i_do_not_exist does not exist!
+
+# Scenario #6: add pk to table containing data invalidating the constraint (this should fail)
+statement ok
+CREATE TABLE test6 (i INTEGER, j INTEGER)
+
+statement ok
+INSERT INTO test6 VALUES (1, 10), (2, 20), (3, 30), (1, 100), (NULL, 0)
+
+statement error
+ALTER TABLE test6 ADD PRIMARY KEY (i)
+----
+PRIMARY KEY or UNIQUE constraint violated: duplicate key "1"
```
Let's divide this into two separate scenarios:
- Creation failure due to duplicate data.
- Creation failure due to `NULL`.

taniabogatsch replied:
Test added.

taniabogatsch commented on test/sql/alter/alter_col/test_add_primary_key.test:
Resolved. Most tests have been added; I'll add separate comments for a few remaining tests.

taniabogatsch commented on src/storage/data_table.cpp:
```diff
@@ -1437,4 +1485,72 @@ vector<ColumnSegmentInfo> DataTable::GetColumnSegmentInfo() {
+	return info->indexes.AddIndex(std::move(art));
+}
+
+void DataTable::IndexExistingData(ClientContext &context, Index &index) {
```
Resolved.

taniabogatsch commented:
Hi @frapa - thanks for adding and implementing most of my comments. I just went over the remaining comments of my previous PR. Could you please add the remaining tests? Once we have all the necessary tests (and their functionality) in this PR, I think we can have another pass about the code and eventually wrap it up.

taniabogatsch commented:
Hi @frapa - are you still working on this? Let me know, and if not, then I can pick it up from here and wrap it up. :)

taniabogatsch commented:
I've picked this up here: https://github.com/taniabogatsch/duckdb/tree/add-pk

taniabogatsch commented:
Hi @frapa - I opened a follow-up PR to finalise your changes and get the feature in: https://github.com/duckdb/duckdb/pull/14419
Thank you for all your work on this!
